graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -578,6 +590,167 @@ def prune_output_blocks_in_compound_match_query(compound_match_query):
+        for match_traversal in match_query.match_traversals:
+            new_match_traversal = []
+            for match_step in match_traversal:
+                # Apply all filters for a location to the first occurence of that location
+                if match_step.as_block.location in location_to_predicates:
+                    where_block = Filter(
+                        _predicate_list_to_where_block(
+                            location_to_predicates[match_step.as_block.location]
+                        )
+                    )
+                    # Delete the location entry. No further filters needed for this location.
+                    del location_to_predicates[match_step.as_block.location]
+                else:
+                    where_block = None
+                new_match_step = MatchStep(
+                    root_block=match_step.root_block,
+                    coerce_type_block=match_step.coerce_type_block,
+                    where_block=where_block,
+                    as_block=match_step.as_block
+                )
+                new_match_traversal.append(new_match_step)
+            new_match_traversals.append(new_match_traversal)
+        new_match_queries.append(
+            MatchQuery(
+                match_traversals=new_match_traversals,
+                folds=match_query.folds,
+                output_block=match_query.output_block
+            )
+        )
+
+    return CompoundMatchQuery(match_queries=new_match_queries)
+
+
+def _update_tagged_expression(expression, current_locations):
+    """Replace non-existent tag Expressons with True, and simplify the result."""
+    if isinstance(expression, BinaryComposition):
+        if isinstance(expression.left, ContextField):
+            context_field = expression.left
+        elif isinstance(expression.right, ContextField):
+            context_field = expression.right
+        elif expression.operator == u'||':
+            if expression.left == TrueLiteral or expression.right == TrueLiteral:
+                return TrueLiteral
+            else:
+                return expression
+        elif expression.operator == u'&&':
+            if expression.left == TrueLiteral and expression.right == TrueLiteral:
+                return TrueLiteral
+            if expression.left == TrueLiteral:
+                return expression.right
+            if expression.right == TrueLiteral:
+                return expression.left
+            else:
+                return expression
+        else:
+            return expression
+        location_name, _ = context_field.location.get_location_name()
+        if location_name not in current_locations:
+            return TrueLiteral
+        else:
+            return expression
+    elif isinstance(expression, TernaryConditional):
+        if expression.predicate == TrueLiteral:
+            return expression.if_true
+        else:
+            return expression
+    elif isinstance(expression, Filter):
```

Comment:
`Filter` is a `Block`, and not an `Expression`. This looks like a type confusion bug.